File: packages/domain/src/resolution/normalization.ts
/**
 * Deterministic alias normalization, fuzzy comparison, and US address parsing.
 */
import type { ParsedAddress } from './types.js';

const ORGANIZATION_SUFFIXES = new Set([
  'association',
  'company',
  'corporation',
  'inc',
  'incorporated',
  'llc',
  'organization',
  'society',
]);

export function normalizeAlias(value: string): string {
  return value
    .normalize('NFKD')
    .replace(/\p{Diacritic}/gu, '')
    .toLocaleLowerCase('en-US')
    .replace(/&/g, ' and ')
    .replace(/['’]/g, '')
    .replace(/[^\p{Letter}\p{Number}]+/gu, ' ')
    .trim()
    .replace(/\s+/g, ' ');
}

export function normalizeOrganizationName(value: string): string {
  const tokens = normalizeAlias(value).split(' ').filter(Boolean);
  while (tokens.length > 1 && ORGANIZATION_SUFFIXES.has(tokens.at(-1)!)) {
    tokens.pop();
  }
  return tokens.join(' ');
}

function levenshtein(left: string, right: string): number {
  if (left === right) return 0;
  if (!left.length) return right.length;
  if (!right.length) return left.length;
  let previous = Array.from({ length: right.length + 1 }, (_, index) => index);
  for (let row = 1; row <= left.length; row += 1) {
    const current = [row];
    for (let column = 1; column <= right.length; column += 1) {
      current[column] = Math.min(
        current[column - 1]! + 1,
        previous[column]! + 1,
        previous[column - 1]! + (left[row - 1] === right[column - 1] ? 0 : 1),
      );
    }
    previous = current;
  }
  return previous[right.length]!;
}

export function nameSimilarity(left: string, right: string): number {
  const normalizedLeft = normalizeAlias(left);
  const normalizedRight = normalizeAlias(right);
  const length = Math.max(normalizedLeft.length, normalizedRight.length);
  if (length === 0) return 1;
  return Math.max(0, 1 - levenshtein(normalizedLeft, normalizedRight) / length);
}

export function parseAddress(raw: string): ParsedAddress {
  const parts = raw
    .split(',')
    .map((part) => part.trim())
    .filter(Boolean);
  const finalPart = parts.at(-1) ?? '';
  const stateZip = finalPart.match(/^([A-Za-z]{2})(?:\s+(\d{5}(?:-\d{4})?))?$/);
  const postalOnly = finalPart.match(/^(\d{5}(?:-\d{4})?)$/);
  const hasState = stateZip !== null;
  const hasPostalOnly = postalOnly !== null;
  const localityEnd = hasState || hasPostalOnly ? parts.length - 1 : parts.length;
  const street = parts.length > 1 ? parts[0] : undefined;
  const city = localityEnd > 1 ? parts[localityEnd - 1] : undefined;
  return {
    raw,
    ...(street ? { street } : {}),
    ...(city ? { city } : {}),
    ...(stateZip?.[1] ? { state: stateZip[1].toUpperCase() } : {}),
    ...(stateZip?.[2]
      ? { postalCode: stateZip[2] }
      : postalOnly?.[1]
        ? { postalCode: postalOnly[1] }
        : {}),
    countryCode: 'US',
  };
}

File: workers/research/src/black_book_research/resolution/normalization.py
"""Deterministic text normalization, fuzzy matching, and US address parsing."""

from __future__ import annotations

import re
import unicodedata


def normalize_alias(value: str) -> str:
    decomposed = unicodedata.normalize("NFKD", value)
    without_marks = "".join(
        char for char in decomposed if not unicodedata.combining(char)
    )
    normalized = (
        without_marks.lower().replace("&", " and ").replace("’", "").replace("'", "")
    )
    return " ".join(re.sub(r"[^\w]+", " ", normalized, flags=re.UNICODE).split())


def _levenshtein(left: str, right: str) -> int:
    previous = list(range(len(right) + 1))
    for row, left_char in enumerate(left, start=1):
        current = [row]
        for column, right_char in enumerate(right, start=1):
            current.append(
                min(
                    current[-1] + 1,
                    previous[column] + 1,
                    previous[column - 1] + (left_char != right_char),
                )
            )
        previous = current
    return previous[-1]


def name_similarity(left: str, right: str) -> float:
    normalized_left = normalize_alias(left)
    normalized_right = normalize_alias(right)
    length = max(len(normalized_left), len(normalized_right))
    return (
        1.0
        if length == 0
        else max(0.0, 1 - _levenshtein(normalized_left, normalized_right) / length)
    )


def parse_address(raw: str) -> dict[str, str]:
    parts = [part.strip() for part in raw.split(",") if part.strip()]
    final = parts[-1] if parts else ""
    state_zip = re.fullmatch(r"([A-Za-z]{2})(?:\s+(\d{5}(?:-\d{4})?))?", final)
    postal_only = re.fullmatch(r"(\d{5}(?:-\d{4})?)", final)
    locality_end = len(parts) - 1 if state_zip or postal_only else len(parts)
    parsed = {"raw": raw, "country_code": "US"}
    if len(parts) > 1:
        parsed["street"] = parts[0]
    if locality_end > 1:
        parsed["city"] = parts[locality_end - 1]
    if state_zip:
        parsed["state"] = state_zip.group(1).upper()
        if state_zip.group(2):
            parsed["postal_code"] = state_zip.group(2)
    elif postal_only:
        parsed["postal_code"] = postal_only.group(1)
    return parsed
